[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.4941748.svg)](https://doi.org/10.5281/zenodo.4941748)

This repo contains the QIIME2 and R code for original research published in _Access Microbiology_:
https://www.microbiologyresearch.org/content/journal/acmi/10.1099/acmi.0.000314

# Antibiotics reduce bacterial load in _Exaiptasia diaphana_, but biofilms hinder its development as a gnotobiotic coral model

## Abstract

Coral reefs are declining due to anthropogenic disturbances, including climate change. Therefore, improving our understanding of coral ecosystems is vital, and the influence of bacteria on coral health has attracted particular interest. However, a gnotobiotic coral model that could enhance studies of coral–bacteria interactions is absent. To address this gap, we tested the ability of treatment with seven antibiotics for 3 weeks to deplete bacteria in _Exaiptasia diaphana_, a sea anemone widely used as a coral model. Digital droplet PCR (ddPCR) targeting anemone _Ef1-α_ and bacterial 16S rRNA genes was used to quantify bacterial load, which was found to decrease six-fold. However, metabarcoding of bacterial 16S rRNA genes showed that alpha and beta diversity of the anemone-associated bacterial communities increased significantly. Therefore, gnotobiotic _E. diaphana_ with simplified, uniform bacterial communities were not generated, with biofilm formation in the culture vessels most likely impeding efforts to eliminate bacteria. Despite this outcome, our work will inform future efforts to create a much needed gnotobiotic coral model.

## Raw Data Availability

The raw sequencing data for this study is deposited under NCBI SRA: PRJNA698456. Raw sequencing data from Maire et al. (2021) was also used in this study and can be found under NCBI SRA: PRJNA650221.
